src/main/java/thePackmaster/patches/MetricsPatches.java

```diff
@@ -0,0 +1,110 @@
+import com.megacrit.cardcrawl.core.Settings;
+import com.megacrit.cardcrawl.dungeons.AbstractDungeon;
+import com.megacrit.cardcrawl.metrics.Metrics;
+import com.megacrit.cardcrawl.monsters.MonsterGroup;
+import com.megacrit.cardcrawl.screens.GameOverScreen;
+import javassist.CtBehavior;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import thePackmaster.SpireAnniversary5Mod;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.stream.Collectors;
+
+import static thePackmaster.ThePackmaster.Enums.THE_PACKMASTER;
+
+public class MetricsPatches {
+    private static final String PMMetricsURL = "http://metrics.twentyfourfuckingletters.com/";
+    private static final Logger logger = LogManager.getLogger(MetricsPatches.class);
+
+    public static ArrayList<String> packChoices = new ArrayList<>();
+
+    @SpirePatch(clz = Metrics.class, method = "sendPost", paramtypez = {String.class, String.class})
+    public static class SendPutInsteadOfPostPatch {
+        @SpireInsertPatch(locator = Locator.class, localvars = "httpRequest")
+        public static void Insert(Metrics metrics, String url, String fileName, Net.HttpRequest httpRequest) {
+            if (AbstractDungeon.player.chosenClass == THE_PACKMASTER) {
+                httpRequest.setMethod("PUT");
+            }
+        }
+
+        private static class Locator extends SpireInsertLocator {
+            @Override
+            public int[] Locate(CtBehavior method) throws Exception {
+                Matcher matcher = new Matcher.MethodCallMatcher(Net.HttpRequest.class, "setContent");
+                return LineFinder.findInOrder(method, matcher);
+            }
+        }
+    }
+
+
+    @SpirePatch(clz = GameOverScreen.class, method = "shouldUploadMetricData")
+    public static class ShouldUploadMetricData {
+        @SpirePostfixPatch
+        public static boolean patch(boolean returnValue) {
+            if (AbstractDungeon.player.chosenClass == THE_PACKMASTER) {
+                returnValue = Settings.UPLOAD_DATA;
+            }
+            return returnValue;
+        }
+
+    }
+
+    @SpirePatch(clz = Metrics.class, method = "run")
+    public static class RunPatch {
+        @SpirePostfixPatch
+        public static void patch(Metrics metrics) {
+            if (metrics.type == Metrics.MetricRequestType.UPLOAD_METRICS && AbstractDungeon.player.chosenClass == THE_PACKMASTER) {
+                try {
+                    Method addDataMethod = Metrics.class.getDeclaredMethod("addData", Object.class, Object.class);
+                    addDataMethod.setAccessible(true);
+                    addDataMethod.invoke(metrics, "pmversion", findTheModVersion());
+                    addDataMethod.invoke(metrics, "currentPacks", SpireAnniversary5Mod.currentPoolPacks.stream().map(p -> p.packID).collect(Collectors.joining(",")));
```

Comment:
Do we want to do anything special to detect all packs mode? Right now it will just record the list of all 90 packs (or however many exist for the player).

It's easy enough to look at the data and work out when all packs mode was on, so I don't have a strong opinion here.

Reply:
Now that I started thinking about all packs mode, another thing we could log is the other config stuff. There are four things you can change from the config menu itself, and there's also the pack summary checkbox on the pack selection screen.
```
SpireAnniversary5Mod.allPacksMode
SpireAnniversary5Mod.oneFrameMode
SpireAnniversary5Mod.sharedContentMode
SpireAnniversary5Mod.allowMultiNone()
SpireAnniversary5Mod.showPackSummaries()
```

Some of these aren't very interesting, but there's some value in knowing how often they're changed. We'd have empirical data about how often people actually changed stuff from the mod settings panel, at least.

This is another thing where I don't have a strong opinion, just doing my best to highlight data that may be worth logging so that we don't unintentionally leave out something that would have been valuable.

Reply:
Pack summaries might be worth logging, good point. Although, since it's toggleable on the pack selection screen as well, it won't really tell us about the mod settings.

The others just aren't very interesting or worth logging in my opinion. Multi none can be easily seen in the custom run config log since you would see multiple nones there. I don#t even know what shared content mode is tbh, probably shares relics or something? Wonder when that came in

Reply:
I agree, most of the settings aren't very interesting, and the interesting ones we can see the effects of by looking at other stuff we're already logging.

Some of the Packmaster relics were initially set up to be available for all characters. I believe MichaelMayhem liked the idea but since it's strongly disliked by many people we changed the relics to be Packmaster only, but added the shared content toggle so that you could opt in to them showing up on other characters. We probably had a brief discussion about it at some point during the initial development phase.